### PR TITLE
Faster Integer Range Operators

### PR DESCRIPTION
**Pull request by manofstick**
As discussed in #520, the current implementation of the range operators is kind of slow. The discussion on #520 listed some issues with the provided implementation which included some issues where the resultant series didn't match, as well as not matching the exceptions thrown in the invalid state of Current (although MSDN docs says that such calls are undefined, so could probably return anything anyway...) These issues, I believe, have been addressed in this implementation.

### CODE REVIEW

**Comment by forki** [comment]
@manofstick do you have any numbers that show that your implementation is indeed faster?

Do we need to add additional tests?

**Comment by manofstick** [comment]
Some performance numbers:

    let sw = System.Diagnostics.Stopwatch.StartNew ()

    let data = seq {
        let r = System.Random 42
        for i = 1 to 10000000 do
            yield r.NextDouble ()    
    }

    let average = data |> Seq.average
    let minimum = data |> Seq.min
    let maximum = data |> Seq.max

    System.Console.WriteLine ("{0} ({1},{2},{3})", sw.ElapsedMilliseconds, minimum, average, maximum)

State | x32 | % of release | x64 | % of release
----- | --- | ------------ | --- | ------------
release | 4664| 100% | 2166| 100%
branch | 1929| 42% | 1278| 59%

TIme in milliseconds

**Comment by manofstick** [comment]
@forki 

This implementation is basically equivalent to @tpetricek's version, (especially if you get rid of the exceptions in Current property, although that was one of the comments that @latkin made - i.e. to match existing functionality)

I also have a "fast path" for the standard increment by 1, which is faster still, but really we're just tweaking a little bit; although it does affect the performance in the previously listed example  by > 10%, so I thought it was worth it.

I probably need to do some better digging to see what existing tests there are to determine if I need to add some, (and probably do; hopefully I'll find some time over the w/end)

**Comment by enricosada** [comment]
@manofstick if you can rebase (or cherry-pick) and force push to remove the merge master commit 9344c31,  it's more clean. Otherwise np

**Comment by forki** [comment]
would it help if the implementation would be special cased for every (or some important like int32) type?

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 1]
```diff
@@ -6497,6 +6497,95 @@ namespace Microsoft.FSharp.Core
...
+            [<NoEquality; NoComparison>]
+            type VariableStepIntegralRangeState<'a> = {
+                mutable Started  : bool
+                mutable Complete : bool
+                mutable Current  : 'a
+            }
+            let inline variableStepIntegralRange n step m =
+                if step = LanguagePrimitives.GenericZero
+                then invalidArg "step" (SR.GetString(SR.stepCannotBeZero));
+                else
```
Comment (enricosada) [inline src/fsharp/FSharp.Core/prim-types.fs]: i think the `else` it's not needed, because `invalidArg` raise an exception ( so a `if then` it's enough, like [there](https://github.com/manofstick/visualfsharp/blob/d5d4c49a30d924ff5388f11eb82cc770e8ca5839/src/fsharp/FSharp.Core/prim-types.fs#L6632)
Less indentation

**Comment by manofstick** [comment]
@enricosada

We use Mercurial at work and so I have never bothered to learn git. Send me commands, and I'll do 'em (...I'm just a noob living in GitHub Desktop land...)

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 2]
```diff
@@ -6497,6 +6497,95 @@ namespace Microsoft.FSharp.Core
...
+                                // is undefined prior to the first call of MoveNext and post called to MoveNext
+                                // that return false (see https://msdn.microsoft.com/en-us/library/58e146b7%28v=vs.110%29.aspx)
+                                // so we should be able to just return value here, and we could get rid of the 
+                                // complete variable which would be faster
+                                if not state.Started then notStarted ()
+                                elif state.Complete  then alreadyFinished ()
+                                else state.Current
+
+                            member this.Current =
+                                box this.Current
```
Comment (enricosada) [inline src/fsharp/FSharp.Core/prim-types.fs]: why `box`? `this.Current` it' not of generic `IEnumerator<'a>` so `abstract Current : 'a with get`?

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 3]
```diff
@@ -6497,6 +6497,95 @@ namespace Microsoft.FSharp.Core
...
+                                      || (step < LanguagePrimitives.GenericZero && next < state.Current && next >= m)
+                                    then
+                                        state.Current <- next
+                                    else
+                                        state.Complete <- true
+
+                                not state.Complete }
+
+                    { new System.Collections.Generic.IEnumerable<'a> with
+                        member __.  GetEnumerator () = variableStepRangeEnumerator ()
```
Comment (enricosada) [inline src/fsharp/FSharp.Core/prim-types.fs]: why `__` and `this` in same class? choose one?

**Comment by manofstick** [comment]
@forki

Due to the statically resolved types, and the way the code is structured (i.e. I do a check for overflow combined with a check for exceeding upper bound) I believe that the code is as efficient as it could be for all data sizes [it is only a replace for "integer" types; i.e not floating point] (... well it could be split for positive and negative step so that check only occurs once, but as I'm dealing with the increment by 1 as a separate case, which I think is the most common usage, I wasn't too concerned)

File: src/fsharp/FSharp.Core/prim-types.fs (hunk 3 above)
Comment (enricosada) [inline src/fsharp/FSharp.Core/prim-types.fs]: i see it's like that in other class in same file. But that's needed or a copy/paste error?

Comment (manofstick) [inline src/fsharp/FSharp.Core/prim-types.fs]: Oh, just a habit where I use __ to denote not using the variable and 'this' when I am. (as usually I'm trying not to use the public interface, so it helps me note where I am).

Happy to follow any convention though.

File: src/fsharp/FSharp.Core/prim-types.fs (hunk 2 above)
Comment (manofstick) [inline src/fsharp/FSharp.Core/prim-types.fs]: It's the "IEnumerable.Current : obj" version.

Probably a bad habbit, but I find myself removing all explicit types.

Happy to put the ": obj" in if it makes this more explicit.

File: src/fsharp/FSharp.Core/prim-types.fs (hunk 3 above)
Comment (enricosada) [inline src/fsharp/FSharp.Core/prim-types.fs]: The golden rule it's use same convention of near code, so you are right, because the usage it's a bit mixed (`x.`, `this.x`, `__.` ) in the file.
Dont know if there is a naming convention or not. If there is we can write that down as a quick note in Docs directory /cc @dsyme.
I think your habit is right, it's nice to have ( i'll use it in my code :smile: )

File: src/fsharp/FSharp.Core/prim-types.fs (hunk 2 above)
Comment (enricosada) [inline src/fsharp/FSharp.Core/prim-types.fs]: ah ok, my bad, it's ok

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 4]
```diff
@@ -6497,6 +6497,95 @@ namespace Microsoft.FSharp.Core
                   interface IEnumerable with 
                       member x.GetEnumerator() = (gen() :> IEnumerator) }
 
+            [<NoEquality; NoComparison>]
+            type VariableStepIntegralRangeState<'a> = {
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/prim-types.fs]: Can this be a struct to save an allocation?

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 5]
```diff
@@ -6497,6 +6497,95 @@ namespace Microsoft.FSharp.Core
...
+                                if   (step > LanguagePrimitives.GenericZero && next > state.Current && next <= m)
+                                    || (step < LanguagePrimitives.GenericZero && next < state.Current && next >= m)
+                                then
+                                    state.Current <- next
+                                else
+                                    state.Complete <- true
+
+                            not state.Complete }
+
+                { new System.Collections.Generic.IEnumerable<'a> with
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/prim-types.fs]: Please use 'T not 'a

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 6]
```diff
@@ -6497,6 +6497,95 @@ namespace Microsoft.FSharp.Core
...
+                    member __.  GetEnumerator () = variableStepRangeEnumerator ()
+                    member this.GetEnumerator () = this.GetEnumerator () :> System.Collections.IEnumerator }
+
+            let inline simpleIntegralRange minValue maxValue n step m =
+                if step <> LanguagePrimitives.GenericOne || n > m || n = minValue || m = maxValue
+                then variableStepIntegralRange n step m
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/prim-types.fs]: Please use

    if ... then 
        expr
    else
        expr

in the F# codebase, thanks

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 7]
```diff
@@ -6497,6 +6497,95 @@ namespace Microsoft.FSharp.Core
...
+                            member __.Dispose () = ()
+                            member __.Current =
+                                // according to IEnumerator<int>.Current documentation, the result of of Current
+                                // is undefined prior to the first call of MoveNext and post called to MoveNext
+                                // that return false (see https://msdn.microsoft.com/en-us/library/58e146b7%28v=vs.110%29.aspx)
+                                // so we should be able to just return value here, which would be faster
+                                if   value < n  then notStarted ()
+                                elif value > m then alreadyFinished ()
+                                else value
+                            member this.Current = box this.Current
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/prim-types.fs]: Please list the implementation of System.Collections.IEnumerator out separately, for clarity

**Comment by dsyme** [comment]
Great work. Please run all tests :) (though I believe master has failing tests at the moment :( )

**Comment by forki** [comment]
How can master have failing tests? ;-)

**Comment by dsyme** [comment]
@forki I know, I know....  Poor CI not able to run enough tests, we have to crank it up,

File: src/fsharp/FSharp.Core/prim-types.fs (hunk 4 above)
Comment (forki) [inline src/fsharp/FSharp.Core/prim-types.fs]: Or maybe we just use 3 different variables and remove this type altogether.

File: src/fsharp/FSharp.Core/prim-types.fs (hunk 5 above)
Comment (manofstick) [inline src/fsharp/FSharp.Core/prim-types.fs]: Probably it's documented somewhere and so I probably deserve a RTFM, but what then is the convention for multiple generic args? 'T, 'R, 'S, etc.. or 'Tfoo, 'Tbar, 'Tbaz? (I kind of liked the small character alphabet) Anyway, I'll update this is 'T.

File: src/fsharp/FSharp.Core/prim-types.fs (hunk 4 above)
Comment (manofstick) [inline src/fsharp/FSharp.Core/prim-types.fs]: Can't say I like the allocation, but without changes to the underlying F# code generator I don't think this is particularly easy to fix (without a lot of duplicated code per type.) So the problem with converting to a struct, or having the 3 variables separately (although I do like from a semantic perspective the mutable state all grouped together) is that the class that F# generates in a closure from `let mutable blah = ...` is just a ref type, so you are paying for the allocation anyway, it's just hidden. So then I thought about actually creating a real type instead of just using the object expression, but because I'm using statically resolved types functionality I would need an inline type which doesn't exist. So I would have to provide a specific type per underlying integer type which seems a bit excessive.

**Comment by dsyme** [comment]
@manofstick I like this, but I suspect we need to add specific new tests to match the cases in the implementation. For example, I look at cases like this: https://github.com/manofstick/visualfsharp/blob/manofstick-perf-range-enumerators/src/fsharp/FSharp.Core/prim-types.fs#L6665 and I reckon we don't have test coverage for edge cases in floating point enumeration.  

Are there specific tests cases we could add?  Perhaps someone could volunteer to do some hammering on this?

**Comment by latkin** [comment]
I had a partial list of nasty cases here https://github.com/Microsoft/visualfsharp/issues/520#issuecomment-117776572, not sure if those are incorporated.

**Comment by manofstick** [comment]
Must say I was a bit slack here, I did kind of assume that strange bounds checks would have already been included in the original test suite, and this was just a refactoring.

But anyway, when I get some time (he says, scrambling to try and put sometime behind the keyboard on a Sunday evening) I'll either ensure that such tests do already exist, or that I add some official ones.

(I did poke an prod them a bit, obviously, so I would be surprised if I stuffed up. But hey, coming off the back of my failed recursive types, I think I would like to double checking my stuff!!)

**Comment by manofstick** [comment]
@dsyme / @latkin 

Oh, there were no changes to the floating point enumerations... We're you just referring in general to have some extra tests?

**Comment by dsyme** [comment]
@manofstick You're right - ignore my comment on floating point, was looking at existing code rather than your new code.  We can assume there is testing for the existing code.

**Comment by dsyme** [comment]
@manofstick Could you check these integer test cases please linked by @latkin above?

```
testSeq 1 10 (Some(Int32.MaxValue))  // new impl overflows, underflows
testSeq 1 10 (Some(-1))              // old impl returns 0-length seq, new impl underflows and gives huge seq
testSeq 3 1 None                     // old impl gives empty seq for a .. b with a < b
```

**Comment by manofstick** [comment]
OK; I'll dig in and punish edge cases (for all the various integer types). Hopefully find time this weekend.

**Comment by manofstick** [comment]
@dsyme 

So much for me trusting that there would have been tests around edges cases to begin with! And actually unless I'm mistaken, I can't actually find any testing around this functionality at all, beyond being utilized in auxiliary functionality in tests in collection modules. (i.e. can't find testing around the integer edges case that are used in floating point sequences that you mentioned before as an example.)

Anyway, I am rather blind, my wife always grabs the thing that I'm looking for from right in front of me, so maybe I'm missing it.

But I'll proceed as if there is currently no testing on ranges; but let me know if you know that there is. (The one place where there does seem to be a little of this kind of testing in regards to BigIntType).

**Comment by dsyme** [comment]
Great, thanks.  There's some in ``tests\fsharp\core\libtest\test.fsx``, e.g. ``IntegerLoopsWithMinAndMaxIntAndKnownBounds``, ``IntegerLoopsWithMinAndMaxIntAndKnownBoundsGoingDown``, but they only test int32.  Minimally duplicating out all these tests for the other numeric types would make sense, plus systematically testing under different strides and edge cases

**Comment by manofstick** [comment]
The tests added give reasonable coverage (I feel) of original functionality, but as singleStepRangeEnumerator special cases single increments when the range bounds are not max and I so I need to add a couple of extra things around that. Anyway, out of time, as usual. Hopefully this weekend at the latest.

File: src/fsharp/FSharp.Core.Unittests/FSharp.Core/PrimTypes.fs [hunk 8]
```diff
@@ -683,3 +683,115 @@ type UnboxAndOptionStuff() =
...
+[<TestFixture>]
+type RangeTests() =
+    [<Test>] member this.SByte  () = RangeTestsHelpers.signed   System.SByte.MinValue  System.SByte.MaxValue
+    [<Test>] member this.Byte   () = RangeTestsHelpers.unsigned System.Byte.MinValue   System.Byte.MaxValue
+    [<Test>] member this.Int16  () = RangeTestsHelpers.signed   System.Int16.MinValue  System.Int16.MaxValue
+    [<Test>] member this.UInt16 () = RangeTestsHelpers.unsigned System.UInt16.MinValue System.UInt16.MaxValue
+    [<Test>] member this.Int32  () = RangeTestsHelpers.signed   System.Int32.MinValue  System.Int32.MaxValue
+    [<Test>] member this.UInt32 () = RangeTestsHelpers.unsigned System.UInt32.MinValue System.UInt32.MaxValue
+    [<Test>] member this.Int64  () = RangeTestsHelpers.signed   System.Int64.MinValue  System.Int64.MaxValue
+    [<Test>] member this.UInt64 () = RangeTestsHelpers.unsigned System.UInt64.MinValue System.UInt64.MaxValue
```
Comment (dsyme) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core/PrimTypes.fs]: What about IntPtr and UIntPtr? Thanks

**Comment by KevinRansom** [comment]
@manofstick @dsyme .  Hey guys, is this ready to pull?

it looks pretty great to me.

**Comment by dsyme** [comment]
Yes, LGTM

File: src/fsharp/FSharp.Core/prim-types.fs (hunk 3 above)
Comment (philderbeast) [inline src/fsharp/FSharp.Core/prim-types.fs]: Wildcard self identifiers (ignore this) is a planned F# change but __ is as close as we can get until the change is made.

* [Wildcard self identifers](https://fslang.uservoice.com/forums/245727-f-language/suggestions/6499392-wildcard-self-identifiers)
* [Allow wildcard self identifiers](https://visualstudio.uservoice.com/forums/121579-visual-studio/suggestions/3745318-allow-wildcard-self-identifiers)
